cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -0,0 +1,146 @@
+            return _repeat_sequence([cirq.YPowGate(), cirq.YPowGate()], num_idle_moments)
+
+
+def _validate_dd_sequence(dd_sequence: list['cirq.Gate']) -> None:
+    if len(dd_sequence) < 2:
+        raise ValueError('Invalid dynamical decoupling sequence. Expect more than one gates.')
+    matrices = [cirq.unitary(gate) for gate in dd_sequence]
+    product = reduce(np.matmul, matrices)
+    if not np.array_equal(product, np.eye(2)):
+        raise ValueError(
+            "Invalid dynamical decoupling sequence, sequence product doesn't equal" ' identity.'
+        )
+
+
+@value.value_equality
+class DynamicalDecouplingModel:
+    """Dynamical decoupling model that generates dynamical decoupling gate sequences."""
+
+    def __init__(
+        self,
+        schema: Optional[_DynamicalDecouplingSchema] = None,
+        base_dd_sequence: Optional[list['cirq.Gate']] = None,
+    ):
+        if not schema and not base_dd_sequence:
+            raise ValueError(
+                'Specify either schema or base_dd_sequence to construct a valid'
+                ' DynamicalDecouplingModel.'
+            )
+        self.schema = schema
+        self.base_dd_sequence = base_dd_sequence
+        if base_dd_sequence:
+            _validate_dd_sequence(base_dd_sequence)
+
+    def generate_dd_sequence(self, num_idle_moments: int = 2) -> list['cirq.Gate']:
+        """Returns the longest possible dynamical decoupling sequence."""
+        if num_idle_moments <= 0:
+            return []
+        if self.schema:
+            return _generate_dd_sequence_from_schema(self.schema, num_idle_moments)
+        if self.base_dd_sequence:
+            return _repeat_sequence(self.base_dd_sequence, num_idle_moments)
+        return []
+
+    @classmethod
+    def from_schema(cls, schema: str):
+        """Create dynamical decoupling model according to a given schema."""
+        if not schema in _DynamicalDecouplingSchema.__members__:
+            raise ValueError("Invalid schema name.")
+        return cls(schema=_DynamicalDecouplingSchema[schema])
+
+    @classmethod
+    def from_base_dd_sequence(cls, base_dd_sequence: list['cirq.Gate']):
+        """Create dynamical decoupling model according to a base sequence."""
+        return cls(base_dd_sequence=base_dd_sequence)
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        d: Dict[str, Any] = {}
+        if self.schema:
+            d['schema'] = self.schema.name
+        if self.base_dd_sequence:
+            d['base_dd_sequence'] = self.base_dd_sequence
+        return d
+
+    @classmethod
+    def _from_json_dict_(cls, schema=None, base_dd_sequence=None, **kwargs):
+        if schema:
+            return cls(schema=_DynamicalDecouplingSchema[schema])
+        if base_dd_sequence:
+            return cls(base_dd_sequence=base_dd_sequence)
+
+    def _value_equality_values_(self) -> Any:
+        return self.schema, self.base_dd_sequence
+
+
+def add_dynamical_decoupling(
```

Comment:
Might be nice to have an option only to insert DD gates on moments that only have single qubits gates. On the hardware, the calibrations for the gates do not include calibrating single qubits and two qubits gates concurrently.

Reply:
Good idea, tracking this [here](https://github.com/quantumlib/Cirq/issues/6582#issuecomment-2089082481). Feel free to put more thoughts there, thanks!